Trade any asset

linkielink opened the pull request:
This PR includes:
- MP-2389 (Enable trading of any asset)
- MP-2388 (Represent non-whitelisted tokens)
- MP-2391 (Update usePrice hook)
- MP-2387 (Deposit and Withdraw any asset)

This is achieved by:
1. fetching all assets for a chain from a config set endpoint (e.g.: DEX API from Astroport)
2. fetch all asset parameters from the assetParams contract
3. enrich the asset data from the dex with the assetParams and static config data (pythFeedIDs, LP tokens and stables)
4. fetch asset prices from the oracle/pyth
5. combine the oracle data with the combine asset data from step 3


Result:
- All usage of `usePrices` got removed
- The `assets` object holds prices now
- All asset listings got automated
- Depositing, Withdrawing and Trading of any asset is now possible

bobthebuidlr commented:
Feedback on `AssetList` in the trade view:

1. Some assets names (denoms) are not truncated, resulting in strange overflow
![CleanShot 2024-06-04 at 15 51 05@2x](https://github.com/mars-protocol/mars-v2-frontend/assets/34470358/a2c271a6-1f2f-4b7a-90aa-18693145b8a7)

2. Favoriting assets doesn't seem to do anything right now.
3. In this video is visible: https://share.cleanshot.com/HWVRzNRZ
    * The collapsible headers don't seem to interact at all
    * There is a strange horizontal scroll space
    * The header just disappears when scrolling down (this was probably always the case). I don't think it make sense to make the header, X button and search bar just scroll away. I'd say that needs to be static or sticky
4. There is no lazy-loading of the asset list rows. This really needs to be put in place. It's a lot to render 100+ list items + images in one go. You can also see this from the delay it takes to open the `AssetList`
5. In the advanced selector there is no way to see what the current selected asset is. It should IMO be on top of the list and already highlighted or something.
6. When selecting an asset from the list, there is almost a 1 second delay before it is actually selected / confirmed.
7. Even when margin trading is disabled, there is a warning message displaying "Not enough liquidity for borrowing"
![CleanShot 2024-06-04 at 16 01 42@2x](https://github.com/mars-protocol/mars-v2-frontend/assets/34470358/b9d4e8af-44a5-469c-9998-cb106fc362de)
8. The N/A is not the correct font size (in 2 places on the fllowing screenshots + max button incorrect size)
![CleanShot 2024-06-04 at 16 04 39@2x](https://github.com/mars-protocol/mars-v2-frontend/assets/34470358/6fcdaa3b-9ac8-4ec4-96af-c10f414bbdf0)
![CleanShot 2024-06-04 at 16 05 16@2x](https://github.com/mars-protocol/mars-v2-frontend/assets/34470358/72849442-7492-40ff-abd4-16ec87073de6)
9. Opening the Account details Sidebar is super laggy too.


Don't know if it is because of this PR, but at least the netValue of my account is horribly incorrect. It should probably be NEGATIVE 250K, but we don't handle this scenario (shouldn' happen anyways if liquidations would be happening). But still wanted to bring this to your attention in case it might have more effect

![CleanShot 2024-06-04 at 16 02 57@2x](https://github.com/mars-protocol/mars-v2-frontend/assets/34470358/7fdf9172-a386-4992-8063-f8e1f8ab4e2e)

linkielink commented:
Fixed the issues mentioned in the comments.

1. Asset Names are now truncated if they use a very long string, that does not contain a slash `/` (like ibc or factory tokens do).
2. Favoriting assets now puts them on top of the list.
3. The collapsible headers were fixed, the horizontal scroll space was removed, the headers don't disaper on scrolling anymore and the closing X stays visible.
4. The asset rows are now sitting under a suspense, while the images (which were indeed set to lazy loading) will load when they come into view. The opening of the AssetsList should now be more performant.
5. Added an `isActive` state to currently selected assets, asset pairs and perps assets.
6. The delay got reduced significantly. Currently the liquidiation price logic spams the memory of the browser, so it can be that this delay is related. But the liquidation price logic will be adjusted in a separate PR.
7. If Margin is not enabled there will be no Low Liquidity warning anymore.
8. Adjusted the Text component to not default to `text-base` but use the text size the parent has, if none is set.
9. Related to the liquidation price and not introduced by the PR.
10. Not related to this PR. The net value is currently set to absolute, this will be corrected in the next PR.

bobthebuidlr commented:
Another few points:
1. Typing in the search bar is delayed as well. The inputs are lagging often.
2. There is a small gap between the first and second asset (see video)
3. When searching for usd and selecting the sell asset, it's possible that there is a strange empty space between the "buy" and "sell" headers

https://share.cleanshot.com/7fcsb5nm

bobthebuidlr commented on src/hooks/assets/useBasetAsset.ts:
```diff
@@ -1,8 +1,12 @@
 
-  return assets[0]
+  return (
+    assets.find((asset) => asset.denom === chainConfig.defaultCurrency.coinMinimalDenom) ??
```
`useMemo`

bobthebuidlr commented on src/hooks/assets/useDisplayCurrencyAssets.ts:
```diff
@@ -1,7 +1,7 @@
-  const assets = useAllAssets()
+  const { data: assets } = useAssets()
 
   return assets.filter((asset) => asset.isDisplayCurrency)
```
`useMemo`

linkielink commented:
Updated according to your feedback. The typing and all other performance issues are most likely related to the liquidation price logic, that I will tackle next.
